packages/editor/src/store/reducer.js

```diff
@@ -361,12 +436,74 @@ export const editor = flow( [
 		byClientId( state = {}, action ) {
 			switch ( action.type ) {
 				case 'SETUP_EDITOR_STATE':
-					return getFlattenedBlocks( action.blocks );
+					return getFlattenedBlocksWithoutAttributes( action.blocks );
+
+				case 'RECEIVE_BLOCKS':
+					return {
+						...state,
+						...getFlattenedBlocksWithoutAttributes( action.blocks ),
+					};
+
+				case 'UPDATE_BLOCK':
+					// Ignore updates if block isn't known
+					if ( ! state[ action.clientId ] ) {
+						return state;
+					}
+
+					return {
+						...state,
+						[ action.clientId ]: {
+							...state[ action.clientId ],
+							...omit( action.updates, 'attributes' ),
+						},
+					};
+
+				case 'INSERT_BLOCKS':
+					return {
+						...state,
+						...getFlattenedBlocksWithoutAttributes( action.blocks ),
+					};
+
+				case 'REPLACE_BLOCKS':
+					if ( ! action.blocks ) {
+						return state;
+					}
+
+					return {
+						...omit( state, action.clientIds ),
+						...getFlattenedBlocksWithoutAttributes( action.blocks ),
+					};
+
+				case 'REMOVE_BLOCKS':
+					return omit( state, action.clientIds );
+			}
+
+			return state;
+		},
+
+		attributesByClientId( state = {}, action ) {
+			switch ( action.type ) {
+				case 'SETUP_EDITOR_STATE':
+					return getFlattenedBlockAttributes( action.blocks );
 
 				case 'RECEIVE_BLOCKS':
 					return {
 						...state,
-						...getFlattenedBlocks( action.blocks ),
+						...getFlattenedBlockAttributes( action.blocks ),
+					};
+
+				case 'UPDATE_BLOCK':
+					// Ignore updates if block isn't known
+					if ( ! state[ action.clientId ] ) {
+						return state;
+					}
+
+					return {
+						...state,
+						[ action.clientId ]: {
+							...state[ action.clientId ],
+							attributes: { ...action.updates.attributes },
```

Comment:
The `attributes` key is redundant here, we can just remove it and update the state to be `{ [ clientID ]: attributes }` instead of `{ [ clientID ]: { attributes } }`

Reply:
Not a blocker for me.

Reply:
> I opted for maintaining the attributes key as a way of having the same general shape in both `byClientId` and `attributesByClientId`, but I'm not married to that choice.

But they're not meant to be the same sort of thing. The former is a block object, the latter (by its name) is exclusively attributes.

What else are we planning to include here? The spread of `...state[ action.clientId ]` seems to imply that there might be other non-`attribute` properties.

Given the name `attributesByClientId`, I would assume the contents of the object are in-fact attributes.

Reply:
Done in latest commit.